sdks/python/apache_beam/io/aws/s3io.py

```diff
@@ -0,0 +1,604 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""AWS S3 client
+"""
+
+from __future__ import absolute_import
+
+import errno
+import io
+import logging
+import re
+import time
+import traceback
+from builtins import object
+
+from apache_beam.io.aws.clients.s3 import messages
+from apache_beam.io.filesystemio import Downloader
+from apache_beam.io.filesystemio import DownloaderStream
+from apache_beam.io.filesystemio import Uploader
+from apache_beam.io.filesystemio import UploaderStream
+from apache_beam.utils import retry
+
+try:
+  # pylint: disable=wrong-import-order, wrong-import-position
+  # pylint: disable=ungrouped-imports
+  from apache_beam.io.aws.clients.s3 import boto3_client
+  BOTO3_INSTALLED = True
+except ImportError:
+  BOTO3_INSTALLED = False
+
+MAX_BATCH_OPERATION_SIZE = 100
+
+
+def parse_s3_path(s3_path, object_optional=False):
+  """Return the bucket and object names of the given s3:// path."""
+  match = re.match('^s3://([^/]+)/(.*)$', s3_path)
+  if match is None or (match.group(2) == '' and not object_optional):
+    raise ValueError('S3 path must be in the form s3://<bucket>/<object>.')
+  return match.group(1), match.group(2)
+
+
+class S3IO(object):
+  """S3 I/O client."""
+
+  def __init__(self, client=None):
+    if client is not None:
+      self.client = client
+    elif BOTO3_INSTALLED:
+      self.client = boto3_client.Client()
+    else:
+      message = 'AWS dependencies are not installed, and no alternative ' \
+      'client was provided to S3IO.'
+      raise RuntimeError(message)
+
+  def open(self,
+           filename,
+           mode='r',
+           read_buffer_size=16*1024*1024,
+           mime_type='application/octet-stream'):
+    """Open an S3 file path for reading or writing.
+
+    Args:
+      filename (str): S3 file path in the form ``s3://<bucket>/<object>``.
+      mode (str): ``'r'`` for reading or ``'w'`` for writing.
+      read_buffer_size (int): Buffer size to use during read operations.
+      mime_type (str): Mime type to set for write operations.
```

Comment:
`mime_type` is ignored in this function. Maybe AWS always works with byte data? Should we verify that users sre requesting bytes?

Reply:
This is addressed with the most recent changes, so the `mime_type` that the user sets will be reflected in the `ContentType` of the object in S3